Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-209.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 7fd52e18cdbacfbb5f75ce2d281bb3a6b1a6dd80
Containerized:False
Number of Error(s): 7
src-react/components/forms/CopyFile.tsx(6,25): error TS2307: Cannot find module ../../hooks/File or its corresponding type declarations.
src-react/components/forms/DesprayFile.tsx(6,25): error TS2307: Cannot find module ../../hooks/File or its corresponding type declarations.
src-react/components/forms/RenameFile.tsx(6,25): error TS2307: Cannot find module ../../hooks/File or its corresponding type declarations.
src-react/components/forms/ReplicateFile.tsx(6,25): error TS2307: Cannot find module ../../hooks/File or its corresponding type declarations.
make[2]: *** [esp/src/CMakeFiles/eclwatch.dir/build.make:311: esp/src/build/dist/stub.eclwatch.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:10418: esp/src/CMakeFiles/eclwatch.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-08-13_20-42-31
time logs started
Prep time   :     8 sec (00:00:08)
Build time  :   164 sec (00:02:44)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   172 sec (00:02:52)
time logs end

[Build error:](https://gist.github.com/4208498f0be06042baf66e0895922361)
Build log [Build log](https://gist.github.com/HPCCSmoketest/4208498f0be06042baf66e0895922361#file-result-21-08-13-20-39-25-log)

--- a/esp/src/src-react/components/FileDetails.tsx
+++ b/esp/src/src-react/components/FileDetails.tsx
@@ -3,6 +3,7 @@ import { CommandBar, ContextualMenuItemType, ICommandBarItemProps, Pivot, PivotI
 import { SizeMe } from "react-sizeme";
 import nlsHPCC from "src/nlsHPCC";
 import { FileParts } from "./FileParts";
+import * as WsDfu  from "src/WsDfu";
 import * as Utility from "src/Utility";
 import { getStateImageName, IFile } from "src/ESPLogicalFile";
 import { useFile, useDefFile } from "../hooks/file";
@@ -15,6 +16,11 @@ import { ECLSourceEditor, XMLSourceEditor } from "./SourceEditor";
 import { ShortVerticalDivider } from "./Common";
 import { FileDetailsGraph } from "./FileDetailsGraph";
 import { TableGroup } from "./forms/Groups";
+import { CopyFile } from "./forms/CopyFile";
+import { DesprayFile } from "./forms/DesprayFile";
+import { RenameFile } from "./forms/RenameFile";
+import { ReplicateFile } from "./forms/ReplicateFile";
+import { Result } from "./Result";
 import { Queries } from "./Queries";
 import { WorkunitDetails } from "./WorkunitDetails";
 
@@ -38,6 +44,12 @@ export const FileDetails: React.FunctionComponent<FileDetailsProps> = ({
     const [description, setDescription] = React.useState("");
     const [_protected, setProtected] = React.useState(false);
     const [restricted, setRestricted] = React.useState(false);
+    const [canReplicateFlag, setCanReplicateFlag] = React.useState(false);
+    const [replicateFlag, setReplicateFlag] = React.useState(false);
+    const [showCopyFile, setShowCopyFile] = React.useState(false);
+    const [showRenameFile, setShowRenameFile] = React.useState(false);
+    const [showDesprayFile, setShowDesprayFile] = React.useState(false);
+    const [showReplicateFile, setShowReplicateFile] = React.useState(false);
 
     const isDFUWorkunit = file?.Wuid[0] === "D";
     const isProtected = file?.ProtectList?.DFUFileProtect?.length > 0 || false;
@@ -47,7 +59,19 @@ export const FileDetails: React.FunctionComponent<FileDetailsProps> = ({
         setProtected(_protected || isProtected);
         setRestricted(restricted || file?.IsRestricted);
 
-    }, [_protected, description, file?.Description, file?.IsRestricted, isProtected, restricted]);
+        if (file?.DFUFilePartsOnClusters?.DFUFilePartsOnCluster?.length > 0) {
+            let _canReplicate = false;
+            let _replicate = false;
+            file?.DFUFilePartsOnClusters?.DFUFilePartsOnCluster.forEach(part => {
+                _canReplicate = _canReplicate && part.CanReplicate;
+                _replicate = _replicate && part.Replicate;
+            });
+            setCanReplicateFlag(_canReplicate);
+            setReplicateFlag(_replicate);
+        }
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [_protected, description, file?.Description, file?.ProtectList?.DFUFileProtect, file?.IsRestricted, file?.DFUFilePartsOnClusters, isProtected, restricted]);
 
     const canSave = file && (
         description !== file.Description ||
@@ -60,6 +84,12 @@ export const FileDetails: React.FunctionComponent<FileDetailsProps> = ({
             key: "refresh", text: nlsHPCC.Refresh, iconProps: { iconName: "Refresh" },
             onClick: () => {
                 refresh();
+            }
+        },
+        {
+            key: "copyFilename", text: nlsHPCC.CopyLogicalFilename, iconProps: { iconName: "Copy" },
+            onClick: () => {
+                navigator?.clipboard?.writeText(logicalFile);
             }
         },
         { key: "divider_1", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
@@ -75,115 +105,144 @@ export const FileDetails: React.FunctionComponent<FileDetailsProps> = ({
             }
         },
         {
-            key: "copy", text: nlsHPCC.CopyLogicalFilename, iconProps: { iconName: "Copy" },
+            key: "delete", text: nlsHPCC.Delete, iconProps: { iconName: "Delete" }, disabled: !file,
             onClick: () => {
-                navigator?.clipboard?.writeText(logicalFile);
+                if (confirm(nlsHPCC.YouAreAboutToDeleteThisFile)) {
+                    WsDfu.DFUArrayAction([file], "Delete").then(response => {
+                        const actionInfo = response?.DFUArrayActionResponse?.ActionResults?.DFUActionInfo;
+                        if (actionInfo && actionInfo.length && !actionInfo[0].Failed) {
+                            window.history.back();
+                        }
+                    });
+                }
             }
         },
         { key: "divider_2", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
-    ], [_protected, canSave, description, file, logicalFile, refresh, restricted]);
+        {
+            key: "copyFile", text: nlsHPCC.Copy, disabled: !file,
+            onClick: () => setShowCopyFile(true)
+        },
+        {
+            key: "rename", text: nlsHPCC.Rename, disabled: !file,
+            onClick: () => setShowRenameFile(true)
+        },
+        {
+            key: "despray", text: nlsHPCC.Despray, disabled: !file,
+            onClick: () => setShowDesprayFile(true)
+        },
+        { key: "divider_3", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "replicate", text: nlsHPCC.Replicate, disabled: !canReplicateFlag || !replicateFlag,
+            onClick: () => setShowReplicateFile(true)
+        },
+    ], [_protected, canReplicateFlag, canSave, description, file, logicalFile, refresh, replicateFlag, restricted]);
 
     const protectedImage = _protected ? Utility.getImageURL("locked.png") : Utility.getImageURL("unlocked.png");
     const stateImage = Utility.getImageURL(getStateImageName(file as unknown as IFile));
     const compressedImage = file?.IsCompressed ? Utility.getImageURL("compressed.png") : "";
 
-    return <SizeMe monitorHeight>{({ size }) =>
-        <Pivot overflowBehavior="menu" style={{ height: "100%" }} selectedKey={tab} onLinkClick={evt => pushUrl(`/files/${cluster}/${logicalFile}/${evt.props.itemKey}`)}>
-            <PivotItem headerText={nlsHPCC.Summary} itemKey="summary" style={pivotItemStyle(size)}>
-                <ScrollablePane scrollbarVisibility={ScrollbarVisibility.auto}>
-                    <Sticky stickyPosition={StickyPositionType.Header}>
-                        <CommandBar items={buttons} />
-                    </Sticky>
-                    <Sticky stickyPosition={StickyPositionType.Header}>
-                        <div style={{ display: "inline-block" }}>
-                            <h2>
-                                <img src={compressedImage} />&nbsp;
-                                <img src={protectedImage} />&nbsp;
-                                <img src={stateImage} />&nbsp;
-                                {file?.Name}
-                            </h2>
-                        </div>
-                    </Sticky>
-                    <TableGroup fields={{
-                        "Wuid": { label: nlsHPCC.Workunit, type: "link", value: file?.Wuid, href: `#/${isDFUWorkunit ? "dfu" : ""}workunits/${file?.Wuid}`, readonly: true, },
-                        "Owner": { label: nlsHPCC.Owner, type: "string", value: file?.Owner, readonly: true },
-                        "SuperOwner": { label: nlsHPCC.SuperFile, type: "links", links: file?.Superfiles?.DFULogicalFile?.map(row => ({ label: "", type: "link", value: row.Name, href: `#/superfiles/${row.Name}` })) },
-                        "NodeGroup": { label: nlsHPCC.ClusterName, type: "string", value: file?.NodeGroup, readonly: true },
-                        "Description": { label: nlsHPCC.Description, type: "string", value: description },
-                        "JobName": { label: nlsHPCC.JobName, type: "string", value: file?.JobName, readonly: true },
-                        "isProtected": { label: nlsHPCC.Protected, type: "checkbox", value: _protected },
-                        "isRestricted": { label: nlsHPCC.Restricted, type: "checkbox", value: restricted },
-                        "ContentType": { label: nlsHPCC.ContentType, type: "string", value: file?.ContentType, readonly: true },
-                        "KeyType": { label: nlsHPCC.KeyType, type: "string", value: file?.KeyType, readonly: true },
-                        "Filesize": { label: nlsHPCC.FileSize, type: "string", value: file?.Filesize, readonly: true },
-                        "Format": { label: nlsHPCC.Format, type: "string", value: file?.Format, readonly: true },
-                        "IsCompressed": { label: nlsHPCC.IsCompressed, type: "checkbox", value: file?.IsCompressed, readonly: true },
-                        "CompressedFileSizeString": { label: nlsHPCC.CompressedFileSize, type: "string", value: file?.CompressedFileSize ? file?.CompressedFileSize.toString() : "", readonly: true },
-                        "PercentCompressed": { label: nlsHPCC.PercentCompressed, type: "string", value: file?.PercentCompressed, readonly: true },
-                        "Modified": { label: nlsHPCC.Modified, type: "string", value: file?.Modified, readonly: true },
-                        "ExpireDays": { label: nlsHPCC.ExpireDays, type: "string", value: file?.ExpireDays ? file?.ExpireDays.toString() : "", readonly: true },
-                        "Directory": { label: nlsHPCC.Directory, type: "string", value: file?.Dir, readonly: true },
-                        "PathMask": { label: nlsHPCC.PathMask, type: "string", value: file?.PathMask, readonly: true },
-                        "RecordSize": { label: nlsHPCC.RecordSize, type: "string", value: file?.RecordSize, readonly: true },
-                        "RecordCount": { label: nlsHPCC.RecordCount, type: "string", value: file?.RecordCount, readonly: true },
-                        "IsReplicated": { label: nlsHPCC.IsReplicated, type: "checkbox", value: file?.DFUFilePartsOnClusters?.DFUFilePartsOnCluster?.length > 0, readonly: true },
-                        "NumParts": { label: nlsHPCC.FileParts, type: "number", value: file?.NumParts, readonly: true },
-                        "MinSkew": { label: nlsHPCC.MinSkew, type: "string", value: file?.Stat?.MinSkew, readonly: true },
-                        "MaxSkew": { label: nlsHPCC.MaxSkew, type: "string", value: file?.Stat?.MaxSkew, readonly: true },
-                        "MinSkewPart": { label: nlsHPCC.MinSkewPart, type: "string", value: file?.Stat?.MinSkewPart === undefined ? "" : file?.Stat?.MinSkewPart.toString(), readonly: true },
-                        "MaxSkewPart": { label: nlsHPCC.MaxSkewPart, type: "string", value: file?.Stat?.MaxSkewPart === undefined ? "" : file?.Stat?.MaxSkewPart.toString(), readonly: true },
-                    }} onChange={(id, value) => {
-                        switch (id) {
-                            case "Description":
-                                setDescription(value);
-                                break;
-                            case "isProtected":
-                                setProtected(value);
-                                break;
-                            case "isRestricted":
-                                setRestricted(value);
-                                break;
-                        }
-                    }} />
-                </ScrollablePane>
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.Contents} itemKey="Contents" style={pivotItemStyle(size, 0)}>
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.DataPatterns} itemKey="DataPatterns" style={pivotItemStyle(size, 0)}>
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.ECL} itemKey="ECL" style={pivotItemStyle(size, 0)}>
-                <ECLSourceEditor text={file?.Ecl} readonly={true} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.DEF} itemKey="DEF" style={pivotItemStyle(size, 0)}>
-                <XMLSourceEditor text={defFile} readonly={true} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.XML} itemKey="XML" style={pivotItemStyle(size, 0)}>
-                <XMLSourceEditor text={xmlFile} readonly={true} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.Superfiles} itemKey="superfiles" itemCount={file?.Superfiles?.DFULogicalFile.length || 0} style={pivotItemStyle(size, 0)}>
-                <SuperFiles cluster={cluster} logicalFile={logicalFile} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.FileParts} itemKey="FileParts" style={pivotItemStyle(size, 0)}>
-                <FileParts cluster={cluster} logicalFile={logicalFile} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.Queries} itemIcon="Search" itemKey="queries" style={pivotItemStyle(size, 0)}>
-                <Queries filter={{ FileName: logicalFile }} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.Queries} itemKey="Queries" style={pivotItemStyle(size, 0)}>
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.Graphs} itemKey="Graphs" itemCount={file?.Graphs?.ECLGraph?.length} headerButtonProps={{ disabled: isDFUWorkunit }} style={pivotItemStyle(size, 0)}>
-                <FileDetailsGraph cluster={cluster} logicalFile={logicalFile} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.Workunit} itemKey="Workunit" style={pivotItemStyle(size, 0)}>
-                {isDFUWorkunit ? <DojoAdapter widgetClassID="DFUWUDetailsWidget" params={{ Wuid: file?.Wuid }} /> : <WorkunitDetails wuid={file?.Wuid} />}
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.History} itemKey="History" style={pivotItemStyle(size, 0)}>
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.Blooms} itemKey="Blooms" itemCount={file?.Blooms?.DFUFileBloom?.length} style={pivotItemStyle(size, 0)}>
-                <FileBlooms cluster={cluster} logicalFile={logicalFile} />
-            </PivotItem>
-            <PivotItem headerText={nlsHPCC.ProtectBy} itemKey="ProtectBy" style={pivotItemStyle(size, 0)}>
-            </PivotItem>
-        </Pivot>
-    }</SizeMe>;
+    return <>
+        <SizeMe monitorHeight>{({ size }) =>
+            <Pivot overflowBehavior="menu" style={{ height: "100%" }} selectedKey={tab} onLinkClick={evt => pushUrl(`/files/${cluster}/${logicalFile}/${evt.props.itemKey}`)}>
+                <PivotItem headerText={nlsHPCC.Summary} itemKey="summary" style={pivotItemStyle(size)}>
+                    <ScrollablePane scrollbarVisibility={ScrollbarVisibility.auto}>
+                        <Sticky stickyPosition={StickyPositionType.Header}>
+                            <CommandBar items={buttons} />
+                        </Sticky>
+                        <Sticky stickyPosition={StickyPositionType.Header}>
+                            <div style={{ display: "inline-block" }}>
+                                <h2>
+                                    <img src={compressedImage} />&nbsp;
+                                    <img src={protectedImage} />&nbsp;
+                                    <img src={stateImage} />&nbsp;
+                                    {file?.Name}
+                                </h2>
+                            </div>
+                        </Sticky>
+                        <TableGroup fields={{
+                            "Wuid": { label: nlsHPCC.Workunit, type: "link", value: file?.Wuid, href: `#/${isDFUWorkunit ? "dfu" : ""}workunits/${file?.Wuid}`, readonly: true, },
+                            "Owner": { label: nlsHPCC.Owner, type: "string", value: file?.Owner, readonly: true },
+                            "SuperOwner": { label: nlsHPCC.SuperFile, type: "links", links: file?.Superfiles?.DFULogicalFile?.map(row => ({ label: "", type: "link", value: row.Name, href: `#/superfiles/${row.Name}` })) },
+                            "NodeGroup": { label: nlsHPCC.ClusterName, type: "string", value: file?.NodeGroup, readonly: true },
+                            "Description": { label: nlsHPCC.Description, type: "string", value: description },
+                            "JobName": { label: nlsHPCC.JobName, type: "string", value: file?.JobName, readonly: true },
+                            "isProtected": { label: nlsHPCC.Protected, type: "checkbox", value: _protected },
+                            "isRestricted": { label: nlsHPCC.Restricted, type: "checkbox", value: restricted },
+                            "ContentType": { label: nlsHPCC.ContentType, type: "string", value: file?.ContentType, readonly: true },
+                            "KeyType": { label: nlsHPCC.KeyType, type: "string", value: file?.KeyType, readonly: true },
+                            "Filesize": { label: nlsHPCC.FileSize, type: "string", value: file?.Filesize, readonly: true },
+                            "Format": { label: nlsHPCC.Format, type: "string", value: file?.Format, readonly: true },
+                            "IsCompressed": { label: nlsHPCC.IsCompressed, type: "checkbox", value: file?.IsCompressed, readonly: true },
+                            "CompressedFileSizeString": { label: nlsHPCC.CompressedFileSize, type: "string", value: file?.CompressedFileSize ? file?.CompressedFileSize.toString() : "", readonly: true },
+                            "PercentCompressed": { label: nlsHPCC.PercentCompressed, type: "string", value: file?.PercentCompressed, readonly: true },
+                            "Modified": { label: nlsHPCC.Modified, type: "string", value: file?.Modified, readonly: true },
+                            "ExpireDays": { label: nlsHPCC.ExpireDays, type: "string", value: file?.ExpireDays ? file?.ExpireDays.toString() : "", readonly: true },
+                            "Directory": { label: nlsHPCC.Directory, type: "string", value: file?.Dir, readonly: true },
+                            "PathMask": { label: nlsHPCC.PathMask, type: "string", value: file?.PathMask, readonly: true },
+                            "RecordSize": { label: nlsHPCC.RecordSize, type: "string", value: file?.RecordSize, readonly: true },
+                            "RecordCount": { label: nlsHPCC.RecordCount, type: "string", value: file?.RecordCount, readonly: true },
+                            "IsReplicated": { label: nlsHPCC.IsReplicated, type: "checkbox", value: file?.DFUFilePartsOnClusters?.DFUFilePartsOnCluster?.length > 0, readonly: true },
+                            "NumParts": { label: nlsHPCC.FileParts, type: "number", value: file?.NumParts, readonly: true },
+                            "MinSkew": { label: nlsHPCC.MinSkew, type: "string", value: file?.Stat?.MinSkew, readonly: true },
+                            "MaxSkew": { label: nlsHPCC.MaxSkew, type: "string", value: file?.Stat?.MaxSkew, readonly: true },
+                            "MinSkewPart": { label: nlsHPCC.MinSkewPart, type: "string", value: file?.Stat?.MinSkewPart === undefined ? "" : file?.Stat?.MinSkewPart.toString(), readonly: true },
+                            "MaxSkewPart": { label: nlsHPCC.MaxSkewPart, type: "string", value: file?.Stat?.MaxSkewPart === undefined ? "" : file?.Stat?.MaxSkewPart.toString(), readonly: true },
+                        }} onChange={(id, value) => {
+                            switch (id) {
+                                case "Description":
+                                    setDescription(value);
+                                    break;
+                                case "isProtected":
+                                    setProtected(value);
+                                    break;
+                                case "isRestricted":
+                                    setRestricted(value);
+                                    break;
+                            }
+                        }} />
+                    </ScrollablePane>
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.Contents} itemKey="Contents" style={pivotItemStyle(size, 0)}>
+                    <Result cluster={cluster} logicalFile={logicalFile} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.DataPatterns} itemKey="DataPatterns" style={pivotItemStyle(size, 0)}>
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.ECL} itemKey="ECL" style={pivotItemStyle(size, 0)}>
+                    <ECLSourceEditor text={file?.Ecl} readonly={true} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.DEF} itemKey="DEF" style={pivotItemStyle(size, 0)}>
+                    <XMLSourceEditor text={defFile} readonly={true} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.XML} itemKey="XML" style={pivotItemStyle(size, 0)}>
+                    <XMLSourceEditor text={xmlFile} readonly={true} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.Superfiles} itemKey="superfiles" itemCount={file?.Superfiles?.DFULogicalFile.length || 0} style={pivotItemStyle(size, 0)}>
+                    <SuperFiles cluster={cluster} logicalFile={logicalFile} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.FileParts} itemKey="FileParts" style={pivotItemStyle(size, 0)}>
+                    <FileParts cluster={cluster} logicalFile={logicalFile} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.Queries} itemKey="queries" style={pivotItemStyle(size, 0)}>
+                    <Queries filter={{ FileName: logicalFile }} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.Graphs} itemKey="Graphs" itemCount={file?.Graphs?.ECLGraph?.length} headerButtonProps={{ disabled: isDFUWorkunit }} style={pivotItemStyle(size, 0)}>
+                    <FileDetailsGraph cluster={cluster} logicalFile={logicalFile} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.Workunit} itemKey="Workunit" style={pivotItemStyle(size, 0)}>
+                    {isDFUWorkunit ? <DojoAdapter widgetClassID="DFUWUDetailsWidget" params={{ Wuid: file?.Wuid }} /> : <WorkunitDetails wuid={file?.Wuid} />}
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.History} itemKey="History" style={pivotItemStyle(size, 0)}>
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.Blooms} itemKey="Blooms" itemCount={file?.Blooms?.DFUFileBloom?.length} style={pivotItemStyle(size, 0)}>
+                    <FileBlooms cluster={cluster} logicalFile={logicalFile} />
+                </PivotItem>
+                <PivotItem headerText={nlsHPCC.ProtectBy} itemKey="ProtectBy" style={pivotItemStyle(size, 0)}>
+                </PivotItem>
+            </Pivot>
+        }</SizeMe>
+        <CopyFile cluster={cluster} logicalFile={logicalFile} showForm={showCopyFile} setShowForm={setShowCopyFile} />
+        <DesprayFile cluster={cluster} logicalFile={logicalFile} showForm={showDesprayFile} setShowForm={setShowDesprayFile} />
+        <RenameFile cluster={cluster} logicalFile={logicalFile} showForm={showRenameFile} setShowForm={setShowRenameFile} />
+        <ReplicateFile cluster={cluster} logicalFile={logicalFile} showForm={showReplicateFile} setShowForm={setShowReplicateFile} />
+    </>;
 };

--- a/esp/src/src-react/components/Result.tsx
+++ b/esp/src/src-react/components/Result.tsx
@@ -206,6 +206,7 @@ interface ResultProps {
     wuid?: string;
     resultName?: string;
     logicalFile?: string;
+    cluster?: string;
     filter?: { [key: string]: any };
 }
 
@@ -215,6 +216,7 @@ export const Result: React.FunctionComponent<ResultProps> = ({
     wuid,
     resultName,
     logicalFile,
+    cluster,
     filter = emptyFilter
 }) => {
 
@@ -222,7 +224,8 @@ export const Result: React.FunctionComponent<ResultProps> = ({
         .baseUrl("")
         .wuid(wuid)
         .resultName(resultName)
-        .logicalFile(resultName)
+        .nodeGroup(cluster)
+        .logicalFile(logicalFile)
         .pagination(true)
         .pageSize(50) as ResultWidget
     );

--- a/esp/src/src-react/components/SuperFiles.tsx
+++ b/esp/src/src-react/components/SuperFiles.tsx
@@ -84,8 +84,8 @@ export const SuperFiles: React.FunctionComponent<SuperFilesProps> = ({
     }, [selection]);
 
     React.useEffect(() => {
-        if (file) {
-            gridStore.setData(file?.Superfiles?.DFULogicalFile);
+        if (file?.Superfiles?.DFULogicalFile) {
+            gridStore?.setData(file?.Superfiles?.DFULogicalFile);
             refreshTable();
         }
     }, [file, gridStore, refreshTable]);

--- a/esp/src/src-react/components/forms/CopyFile.tsx
+++ b/esp/src/src-react/components/forms/CopyFile.tsx
@@ -1,0 +1,236 @@
+import * as React from "react";
+import { Checkbox, ContextualMenu, IconButton, IDragOptions, mergeStyleSets, Modal, PrimaryButton, Stack, TextField, } from "@fluentui/react";
+import { useId } from "@fluentui/react-hooks";
+import { useForm, Controller } from "react-hook-form";
+import nlsHPCC from "src/nlsHPCC";
+import { useFile } from "../../hooks/file";
+import * as FileSpray from "src/FileSpray";
+import { TargetGroupTextField } from "./Fields";
+import * as FormStyles from "./landing-zone/styles";
+import { pushUrl } from "../../util/history";
+
+interface CopyFileFormValues {
+    destGroup: string;
+    destLogicalName: string;
+    overwrite: boolean;
+    replicate: boolean;
+    nosplit: boolean;
+    compress: boolean;
+    Wrap: boolean;
+    superCopy: boolean;
+    preserveCompression: boolean;
+    ExpireDays: string;
+}
+
+const defaultValues: CopyFileFormValues = {
+    destGroup: "",
+    destLogicalName: "",
+    overwrite: false,
+    replicate: false,
+    nosplit: false,
+    compress: false,
+    Wrap: false,
+    superCopy: false,
+    preserveCompression: true,
+    ExpireDays: ""
+};
+
+interface CopyFileProps {
+    cluster: string;
+    logicalFile: string;
+
+    showForm: boolean;
+    setShowForm: (_: boolean) => void;
+}
+
+export const CopyFile: React.FunctionComponent<CopyFileProps> = ({
+    cluster,
+    logicalFile,
+    showForm,
+    setShowForm
+}) => {
+
+    const [file] = useFile(cluster, logicalFile);
+
+    const { handleSubmit, control, reset } = useForm<CopyFileFormValues>({ defaultValues });
+
+    const closeForm = React.useCallback(() => {
+        setShowForm(false);
+    }, [setShowForm]);
+
+    const onSubmit = React.useCallback(() => {
+        handleSubmit(
+            (data, evt) => {
+                const request = { ...data, sourceLogicalName: logicalFile };
+                FileSpray.Copy({ request: request }).then(response => {
+                    closeForm();
+                    pushUrl(`/files/${data.destGroup}/${data.destLogicalName}`);
+                });
+            },
+            err => {
+                console.log(err);
+            }
+        )();
+    }, [closeForm, handleSubmit, logicalFile]);
+
+    const titleId = useId("title");
+
+    const dragOptions: IDragOptions = {
+        moveMenuItemText: nlsHPCC.Move,
+        closeMenuItemText: nlsHPCC.Close,
+        menu: ContextualMenu,
+    };
+
+    const componentStyles = mergeStyleSets(
+        FormStyles.componentStyles,
+        {
+            container: {
+                minWidth: 440,
+            }
+        }
+    );
+
+    React.useEffect(() => {
+        const newValues = { ...defaultValues, destLogicalName: logicalFile };
+        reset(newValues);
+    }, [file, logicalFile, reset]);
+
+    return <Modal
+        titleAriaId={titleId}
+        isOpen={showForm}
+        onDismiss={closeForm}
+        isBlocking={false}
+        containerClassName={componentStyles.container}
+        dragOptions={dragOptions}
+    >
+        <div className={componentStyles.header}>
+            <span id={titleId}>{nlsHPCC.Copy}</span>
+            <IconButton
+                styles={FormStyles.iconButtonStyles}
+                iconProps={FormStyles.cancelIcon}
+                ariaLabel={nlsHPCC.CloseModal}
+                onClick={closeForm}
+            />
+        </div>
+        <div className={componentStyles.body}>
+            <Stack>
+                <Controller
+                    control={control} name="destGroup"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TargetGroupTextField
+                            key={fieldName}
+                            label={nlsHPCC.Group}
+                            required={true}
+                            selectedKey={value}
+                            placeholder={nlsHPCC.SelectValue}
+                            onChange={(evt, option) => {
+                                onChange(option.key);
+                            }}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: `${nlsHPCC.SelectA} ${nlsHPCC.Group}`
+                    }}
+                />
+                <Controller
+                    control={control} name="destLogicalName"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TextField
+                            name={fieldName}
+                            onChange={onChange}
+                            required={true}
+                            label={nlsHPCC.TargetName}
+                            value={value}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: nlsHPCC.ValidationErrorRequired
+                    }}
+                />
+            </Stack>
+            <Stack>
+                <table className={componentStyles.twoColumnTable}>
+                    <tbody>
+                        <tr>
+                            <td><Controller
+                                control={control} name="overwrite"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.Overwrite} />}
+                            /></td>
+                            <td><Controller
+                                control={control} name="replicate"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.Replicate} />}
+                            /></td>
+                        </tr>
+                        <tr>
+                            <td><Controller
+                                control={control} name="nosplit"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.NoSplit} />}
+                            /></td>
+                            <td><Controller
+                                control={control} name="compress"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.Compress} />}
+                            /></td>
+                        </tr>
+                        <tr>
+                            <td><Controller
+                                control={control} name="Wrap"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.Wrap} />}
+                            /></td>
+                            <td><Controller
+                                control={control} name="superCopy"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.RetainSuperfileStructure} />}
+                            /></td>
+                        </tr>
+                        <tr>
+                            <td><Controller
+                                control={control} name="preserveCompression"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.PreserveCompression} />}
+                            /></td>
+                            <td><Controller
+                                control={control} name="ExpireDays" defaultValue={""}
+                                render={({
+                                    field: { onChange, name: fieldName, value },
+                                    fieldState: { error }
+                                }) => <TextField
+                                        name={fieldName}
+                                        onChange={onChange}
+                                        label={nlsHPCC.ExpireDays}
+                                        value={value}
+                                        errorMessage={error && error.message}
+                                    />}
+                                rules={{
+                                    min: {
+                                        value: 1,
+                                        message: `${nlsHPCC.ValidationErrorNumberLess} 1`
+                                    }
+                                }}
+                            /></td>
+                        </tr>
+                    </tbody>
+                </table>
+            </Stack>
+
+            <Stack horizontal horizontalAlign="space-between" verticalAlign="end" styles={FormStyles.buttonStackStyles}>
+                <PrimaryButton text={nlsHPCC.Copy} onClick={handleSubmit(onSubmit)} />
+            </Stack>
+        </div>
+    </Modal>;
+};

--- a/esp/src/src-react/components/forms/DesprayFile.tsx
+++ b/esp/src/src-react/components/forms/DesprayFile.tsx
@@ -1,0 +1,251 @@
+import * as React from "react";
+import { Checkbox, ContextualMenu, IconButton, IDragOptions, mergeStyleSets, Modal, PrimaryButton, Stack, TextField, } from "@fluentui/react";
+import { useId } from "@fluentui/react-hooks";
+import { useForm, Controller } from "react-hook-form";
+import nlsHPCC from "src/nlsHPCC";
+import { useFile } from "../../hooks/file";
+import * as FileSpray from "src/FileSpray";
+import { TargetDropzoneTextField, TargetFolderTextField, TargetServerTextField } from "./Fields";
+import * as FormStyles from "./landing-zone/styles";
+import { pushUrl } from "../../util/history";
+
+interface DesprayFileFormValues {
+    destGroup: string;
+    destIP: string;
+    destPath: string;
+    targetName: string;
+    splitprefix: string;
+    overwrite: boolean;
+    SingleConnection: boolean;
+    wrap: boolean;
+}
+
+const defaultValues: DesprayFileFormValues = {
+    destGroup: "",
+    destIP: "",
+    destPath: "",
+    targetName: "",
+    splitprefix: "",
+    overwrite: false,
+    SingleConnection: false,
+    wrap: false
+};
+
+interface DesprayFileProps {
+    cluster: string;
+    logicalFile: string;
+
+    showForm: boolean;
+    setShowForm: (_: boolean) => void;
+}
+
+export const DesprayFile: React.FunctionComponent<DesprayFileProps> = ({
+    cluster,
+    logicalFile,
+    showForm,
+    setShowForm
+}) => {
+
+    const [file] = useFile(cluster, logicalFile);
+    const [machine, setMachine] = React.useState<string>("");
+    const [directory, setDirectory] = React.useState<string>("/");
+    const [pathSep, setPathSep] = React.useState<string>("/");
+    const [os, setOs] = React.useState<number>();
+
+    const { handleSubmit, control, reset } = useForm<DesprayFileFormValues>({ defaultValues });
+
+    const closeForm = React.useCallback(() => {
+        setShowForm(false);
+    }, [setShowForm]);
+
+    const onSubmit = React.useCallback(() => {
+        handleSubmit(
+            (data, evt) => {
+                const request = {
+                    ...data,
+                    destPath: [data.destPath, data.targetName].join(pathSep),
+                    sourceLogicalName: logicalFile
+                };
+                FileSpray.Despray({ request: request }).then(response => {
+                    closeForm();
+                    pushUrl(`/dfuworkunits/${response?.DesprayResponse?.wuid}`);
+                });
+            },
+            err => {
+                console.log(err);
+            }
+        )();
+    }, [closeForm, handleSubmit, logicalFile, pathSep]);
+
+    const titleId = useId("title");
+
+    const dragOptions: IDragOptions = {
+        moveMenuItemText: nlsHPCC.Move,
+        closeMenuItemText: nlsHPCC.Close,
+        menu: ContextualMenu,
+    };
+
+    const componentStyles = mergeStyleSets(
+        FormStyles.componentStyles,
+        {
+            container: {
+                minWidth: 440,
+            }
+        }
+    );
+
+    React.useEffect(() => {
+        const newValues = { ...defaultValues, targetName: file?.Filename };
+        reset(newValues);
+    }, [file?.Filename, reset]);
+
+    return <Modal
+        titleAriaId={titleId}
+        isOpen={showForm}
+        onDismiss={closeForm}
+        isBlocking={false}
+        containerClassName={componentStyles.container}
+        dragOptions={dragOptions}
+    >
+        <div className={componentStyles.header}>
+            <span id={titleId}>{nlsHPCC.Despray}</span>
+            <IconButton
+                styles={FormStyles.iconButtonStyles}
+                iconProps={FormStyles.cancelIcon}
+                ariaLabel={nlsHPCC.CloseModal}
+                onClick={closeForm}
+            />
+        </div>
+        <div className={componentStyles.body}>
+            <Stack>
+                <Controller
+                    control={control} name="destGroup"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TargetDropzoneTextField
+                            key={fieldName}
+                            label={nlsHPCC.DropZone}
+                            required={true}
+                            selectedKey={value}
+                            placeholder={nlsHPCC.SelectValue}
+                            onChange={(evt, option) => {
+                                setDirectory(option["path"] as string);
+                                if (option["path"].indexOf("\\") > -1) {
+                                    setPathSep("\\");
+                                }
+                                onChange(option.key);
+                            }}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: `${nlsHPCC.SelectA} ${nlsHPCC.DropZone}`
+                    }}
+                />
+                <Controller
+                    control={control} name="destIP"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TargetServerTextField
+                            key={fieldName}
+                            required={true}
+                            label={nlsHPCC.IPAddress}
+                            selectedKey={value}
+                            placeholder={nlsHPCC.SelectValue}
+                            onChange={(evt, option) => {
+                                setMachine(option.key as string);
+                                setOs(option["OS"] as number);
+                                onChange(option.key);
+                            }}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: nlsHPCC.ValidationErrorRequired
+                    }}
+                />
+                <Controller
+                    control={control} name="destPath"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TargetFolderTextField
+                            key={fieldName}
+                            label={nlsHPCC.Path}
+                            pathSepChar={pathSep}
+                            machineAddress={machine}
+                            machineDirectory={directory}
+                            machineOS={os}
+                            required={true}
+                            placeholder={nlsHPCC.SelectValue}
+                            onChange={(evt, option) => {
+                                onChange(option.key);
+                            }}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: nlsHPCC.ValidationErrorRequired,
+                    }}
+                />
+                <Controller
+                    control={control} name="targetName"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TextField
+                            name={fieldName}
+                            onChange={onChange}
+                            required={true}
+                            label={nlsHPCC.TargetName}
+                            value={value}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: nlsHPCC.ValidationErrorRequired
+                    }}
+                />
+                <Controller
+                    control={control} name="splitprefix"
+                    render={({
+                        field: { onChange, name: fieldName, value }
+                    }) => <TextField
+                            name={fieldName}
+                            onChange={onChange}
+                            label={nlsHPCC.SplitPrefix}
+                            value={value}
+                        />}
+                />
+            </Stack>
+            <Stack>
+                <table className={componentStyles.twoColumnTable}>
+                    <tbody><tr>
+                        <td><Controller
+                            control={control} name="overwrite"
+                            render={({
+                                field: { onChange, name: fieldName, value }
+                            }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.Overwrite} />}
+                        /></td>
+                        <td><Controller
+                            control={control} name="SingleConnection"
+                            render={({
+                                field: { onChange, name: fieldName, value }
+                            }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.UseSingleConnection} />}
+                        /></td>
+                    </tr>
+                        <tr>
+                            <td><Controller
+                                control={control} name="wrap"
+                                render={({
+                                    field: { onChange, name: fieldName, value }
+                                }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.PreserveParts} />}
+                            /></td>
+                        </tr></tbody>
+                </table>
+            </Stack>
+
+            <Stack horizontal horizontalAlign="space-between" verticalAlign="end" styles={FormStyles.buttonStackStyles}>
+                <PrimaryButton text={nlsHPCC.Despray} onClick={handleSubmit(onSubmit)} />
+            </Stack>
+        </div>
+    </Modal>;
+};

--- a/esp/src/src-react/components/forms/Fields.tsx
+++ b/esp/src/src-react/components/forms/Fields.tsx
@@ -708,6 +708,7 @@ export function createInputs(fields: Fields, onChange?: (id: string, newValue: a
                     label: field.label,
                     field: <TargetGroupTextField
                         key={fieldID}
+                        required={field.required}
                         selectedKey={field.value}
                         onChange={(ev, row) => onChange(fieldID, row.key)}
                         placeholder={field.placeholder}

--- a/esp/src/src-react/components/forms/Filter.tsx
+++ b/esp/src/src-react/components/forms/Filter.tsx
@@ -89,7 +89,7 @@ export const Filter: React.FunctionComponent<FilterProps> = ({
             <IconButton
                 styles={iconButtonStyles}
                 iconProps={cancelIcon}
-                ariaLabel="Close popup modal"
+                ariaLabel={nlsHPCC.CloseModal}
                 onClick={closeFilter}
             />
         </div>

--- a/esp/src/src-react/components/forms/RenameFile.tsx
+++ b/esp/src/src-react/components/forms/RenameFile.tsx
@@ -1,0 +1,132 @@
+import * as React from "react";
+import { Checkbox, ContextualMenu, IconButton, IDragOptions, mergeStyleSets, Modal, PrimaryButton, Stack, TextField, } from "@fluentui/react";
+import { useId } from "@fluentui/react-hooks";
+import { useForm, Controller } from "react-hook-form";
+import nlsHPCC from "src/nlsHPCC";
+import { useFile } from "../../hooks/file";
+import * as FileSpray from "src/FileSpray";
+import * as FormStyles from "./landing-zone/styles";
+import { pushUrl } from "../../util/history";
+
+interface RenameFileFormValues {
+    dstname: string;
+    overwrite: boolean;
+}
+
+const defaultValues: RenameFileFormValues = {
+    dstname: "",
+    overwrite: false
+};
+
+interface RenameFileProps {
+    cluster: string;
+    logicalFile: string;
+
+    showForm: boolean;
+    setShowForm: (_: boolean) => void;
+}
+
+export const RenameFile: React.FunctionComponent<RenameFileProps> = ({
+    cluster,
+    logicalFile,
+    showForm,
+    setShowForm
+}) => {
+
+    const [file] = useFile(cluster, logicalFile);
+
+    const { handleSubmit, control, reset } = useForm<RenameFileFormValues>({ defaultValues });
+
+    const closeForm = React.useCallback(() => {
+        setShowForm(false);
+    }, [setShowForm]);
+
+    const onSubmit = React.useCallback(() => {
+        handleSubmit(
+            (data, evt) => {
+                const request = { ...data, srcname: logicalFile };
+                FileSpray.Rename({ request: request }).then(response => {
+                    closeForm();
+                    pushUrl(`/files/${cluster}/${data.dstname}`);
+                });
+            },
+            err => {
+                console.log(err);
+            }
+        )();
+    }, [closeForm, cluster, handleSubmit, logicalFile]);
+
+    const titleId = useId("title");
+
+    const dragOptions: IDragOptions = {
+        moveMenuItemText: nlsHPCC.Move,
+        closeMenuItemText: nlsHPCC.Close,
+        menu: ContextualMenu,
+    };
+
+    const componentStyles = mergeStyleSets(
+        FormStyles.componentStyles,
+        {
+            container: {
+                minWidth: 440,
+            }
+        }
+    );
+
+    React.useEffect(() => {
+        const newValues = { ...defaultValues, dstname: logicalFile };
+        reset(newValues);
+    }, [file, logicalFile, reset]);
+
+    return <Modal
+        titleAriaId={titleId}
+        isOpen={showForm}
+        onDismiss={closeForm}
+        isBlocking={false}
+        containerClassName={componentStyles.container}
+        dragOptions={dragOptions}
+    >
+        <div className={componentStyles.header}>
+            <span id={titleId}>{nlsHPCC.Rename}</span>
+            <IconButton
+                styles={FormStyles.iconButtonStyles}
+                iconProps={FormStyles.cancelIcon}
+                ariaLabel={nlsHPCC.CloseModal}
+                onClick={closeForm}
+            />
+        </div>
+        <div className={componentStyles.body}>
+            <Stack>
+                <Controller
+                    control={control} name="dstname"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TextField
+                            name={fieldName}
+                            onChange={onChange}
+                            required={true}
+                            label={nlsHPCC.TargetName}
+                            value={value}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: nlsHPCC.ValidationErrorRequired
+                    }}
+                />
+                <div style={{ paddingTop: "15px" }}>
+                    <Controller
+                        control={control} name="overwrite"
+                        render={({
+                            field: { onChange, name: fieldName, value }
+                        }) => <Checkbox name={fieldName} checked={value} onChange={onChange} label={nlsHPCC.Overwrite} />}
+                    />
+                </div>
+            </Stack>
+
+            <Stack horizontal horizontalAlign="space-between" verticalAlign="end" styles={FormStyles.buttonStackStyles}>
+                <PrimaryButton text={nlsHPCC.Rename} onClick={handleSubmit(onSubmit)} />
+            </Stack>
+        </div>
+    </Modal>;
+};

--- a/esp/src/src-react/components/forms/ReplicateFile.tsx
+++ b/esp/src/src-react/components/forms/ReplicateFile.tsx
@@ -1,0 +1,165 @@
+import * as React from "react";
+import { ContextualMenu, IconButton, IDragOptions, mergeStyleSets, Modal, PrimaryButton, Stack, TextField, } from "@fluentui/react";
+import { useId } from "@fluentui/react-hooks";
+import { useForm, Controller } from "react-hook-form";
+import nlsHPCC from "src/nlsHPCC";
+import { useFile } from "../../hooks/file";
+import * as FileSpray from "src/FileSpray";
+import { TargetGroupTextField } from "./Fields";
+import * as FormStyles from "./landing-zone/styles";
+import { pushUrl } from "../../util/history";
+
+interface ReplicateFileFormValues {
+    sourceLogicalName: string;
+    replicateOffset: string;
+    cluster: string;
+}
+
+const defaultValues: ReplicateFileFormValues = {
+    sourceLogicalName: "",
+    replicateOffset: "1",
+    cluster: ""
+};
+
+interface ReplicateFileProps {
+    cluster: string;
+    logicalFile: string;
+
+    showForm: boolean;
+    setShowForm: (_: boolean) => void;
+}
+
+export const ReplicateFile: React.FunctionComponent<ReplicateFileProps> = ({
+    cluster,
+    logicalFile,
+    showForm,
+    setShowForm
+}) => {
+
+    const [file] = useFile(cluster, logicalFile);
+    const { handleSubmit, control, reset } = useForm<ReplicateFileFormValues>({ defaultValues });
+
+    const closeForm = React.useCallback(() => {
+        setShowForm(false);
+    }, [setShowForm]);
+
+    const onSubmit = React.useCallback(() => {
+        handleSubmit(
+            (data, evt) => {
+                const request = { ...data, srcname: logicalFile };
+                FileSpray.Replicate({ request: request }).then(response => {
+                    closeForm();
+                    pushUrl(`/dfuworkunits/${response?.ReplicateResponse?.wuid}`);
+                });
+            },
+            err => {
+                console.log(err);
+            }
+        )();
+    }, [closeForm, handleSubmit, logicalFile]);
+
+    const titleId = useId("title");
+
+    const dragOptions: IDragOptions = {
+        moveMenuItemText: nlsHPCC.Move,
+        closeMenuItemText: nlsHPCC.Close,
+        menu: ContextualMenu,
+    };
+
+    const componentStyles = mergeStyleSets(
+        FormStyles.componentStyles,
+        {
+            container: {
+                minWidth: 440,
+            }
+        }
+    );
+
+    React.useEffect(() => {
+        const newValues = { ...defaultValues, sourceLogicalName: logicalFile };
+        reset(newValues);
+    }, [file, logicalFile, reset]);
+
+    return <Modal
+        titleAriaId={titleId}
+        isOpen={showForm}
+        onDismiss={closeForm}
+        isBlocking={false}
+        containerClassName={componentStyles.container}
+        dragOptions={dragOptions}
+    >
+        <div className={componentStyles.header}>
+            <span id={titleId}>{nlsHPCC.Rename}</span>
+            <IconButton
+                styles={FormStyles.iconButtonStyles}
+                iconProps={FormStyles.cancelIcon}
+                ariaLabel={nlsHPCC.CloseModal}
+                onClick={closeForm}
+            />
+        </div>
+        <div className={componentStyles.body}>
+            <Stack>
+                <Controller
+                    control={control} name="sourceLogicalName"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TextField
+                            name={fieldName}
+                            onChange={onChange}
+                            required={true}
+                            label={nlsHPCC.SourceLogicalFile}
+                            value={value}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: nlsHPCC.ValidationErrorRequired
+                    }}
+                />
+                <Controller
+                    control={control} name="replicateOffset"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TextField
+                            name={fieldName}
+                            onChange={onChange}
+                            label={nlsHPCC.ReplicateOffset}
+                            value={value}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        pattern: {
+                            value: /^[0-9]+$/i,
+                            message: nlsHPCC.ValidationErrorEnterNumber
+                        }
+                    }}
+                />
+                <Controller
+                    control={control} name="cluster"
+                    render={({
+                        field: { onChange, name: fieldName, value },
+                        fieldState: { error }
+                    }) => <TargetGroupTextField
+                            key={fieldName}
+                            label={nlsHPCC.Cluster}
+                            required={true}
+                            selectedKey={value}
+                            placeholder={nlsHPCC.SelectValue}
+                            onChange={(evt, option) => {
+                                onChange(option.key);
+                            }}
+                            errorMessage={error && error.message}
+                        />}
+                    rules={{
+                        required: `${nlsHPCC.SelectA} ${nlsHPCC.Cluster}`
+                    }}
+                />
+            </Stack>
+
+            <Stack horizontal horizontalAlign="space-between" verticalAlign="end" styles={FormStyles.buttonStackStyles}>
+                <PrimaryButton text={nlsHPCC.Replicate} onClick={handleSubmit(onSubmit)} />
+            </Stack>
+        </div>
+    </Modal>;
+};

